pkg/controller/infra/infra_controller.go

```diff
@@ -201,18 +366,145 @@ func (c *InfraController) handleErr(err error, key interface{}) {
 		return
 	}
 
-	c.logger.WithField("cluster", key).Errorf("error syncing cluster: %v", err)
+	logger := c.logger.WithField("cluster", key)
+
+	logger.Errorf("error syncing cluster: %v", err)
 	if c.queue.NumRequeues(key) < maxRetries {
-		c.logger.WithField("cluster", key).Errorf("error syncing cluster: %v", err)
+		logger.Errorf("retrying cluster")
 		c.queue.AddRateLimited(key)
 		return
 	}
 
 	utilruntime.HandleError(err)
-	c.logger.WithField("cluster", key).Infof("dropping cluster out of the queue: %v", err)
+	logger.Infof("dropping cluster out of the queue: %v", err)
 	c.queue.Forget(key)
 }
 
+func (c *InfraController) syncClusterStatusWithJob(cluster *clusteroperator.Cluster, job *v1batch.Job) error {
+	cluster = cluster.DeepCopy()
+	now := metav1.Now()
+
+	// Ensure that job status reflects current job
+	if cluster.Status.ProvisioningJob == nil {
+		cluster.Status.ProvisioningJob = &kapi.LocalObjectReference{
+			Name: job.Name,
+		}
+		cluster.Status.ProvisioningJobGeneration = jobClusterGeneration(job)
```

Comment:
I think we need to set `ProvisioningJobGeneration` when the job is created and use comparison between the cluster generation stored in the job's annotations and the current generation of the cluster to determine whether a job matches up with the current state of the cluster for updating conditions.